router/core/router.go

```diff
@@ -962,7 +980,45 @@ func (r *Router) Start(ctx context.Context) error {
 			return err
 		}
 
-		return r.listenAndServe(r.staticRouterConfig)
+		if err := r.listenAndServe(r.staticRouterConfig); err != nil {
+			return err
+		}
+
+		if r.executionConfig != nil && r.executionConfig.Watch {
+			watcher, err := file_watcher.FileWatcher(r.logger, r.executionConfig.Path)
+			if err != nil {
+				return fmt.Errorf("failed to watch config file: %w", err)
+			}
+			err = watcher.Watch(func() {
+				if r.shutdown.Load() {
+					r.logger.Warn("Router is in shutdown state. Skipping config update")
+				}
+
+				data, err := os.ReadFile(r.executionConfig.Path)
+				if err != nil {
+					r.logger.Error("Failed to read config file", zap.Error(err))
+					return
+				}
+
+				r.logger.Info("Config file changed. Updating server with new config")
+
+				cfg, err := execution_config.SerializeConfigBytes(data)
+				if err != nil {
+					r.logger.Error("Failed to serialize config file", zap.Error(err))
+					return
+				}
+
+				if err := r.newServer(ctx, cfg); err != nil {
```

Comment:
Is it possible to re-use logic from other polling/update mechanisms?

Reply:
We do. The only code we added here is specific to the watcher. The implementation that polls storage providers for updates is already unified.